### PR TITLE
[X86] AMD Zen 5 Initial enablement

### PR DESCRIPTION
**Pull request by ganeshgit**
This patch enables the basic skeleton enablement of AMD next gen zen5 CPUs.

### CODE REVIEW

File: llvm/lib/TargetParser/Host.cpp
Comment (boomanaiden154) [inline llvm/lib/TargetParser/Host.cpp:1154]: Can you bump the equivalent code in `compiler-rt` too?

Comment (ganeshgit) [reply llvm/lib/TargetParser/Host.cpp:1154]: Will do!

Comment (boomanaiden154) [reply llvm/lib/TargetParser/Host.cpp:1154]: I posted some patches a while ago to start unifying things so that there's a single canonical version that gets copied and pasted, but those have been stalled for a while on reviewers, so for now we're mostly stuck with the ad-hoc way of doing things.

Comment (ganeshgit) [reply llvm/lib/TargetParser/Host.cpp:1154]: > I posted some patches a while ago to start unifying things so that there's a single canonical version that gets copied and pasted, but those have been stalled for a while on reviewers, so for now we're mostly stuck with the ad-hoc way of doing things.

Can you point me to those pull requests.
BTW, I see that the compiler-rt change is in place. Do you find any issues there? I think the ordering with respect to libgcc is also okay. 
https://github.com/llvm/llvm-project/pull/107964/commits/b68bcc1415151bd84b5868aa2c98663069f45469#diff-f0c447f4acaa87cede210a02796f8a3b2f08f6d8acaf951ed4c65465bf37c967

Comment (boomanaiden154) [reply llvm/lib/TargetParser/Host.cpp:1154]: Oh, looks like I missed it. Sorry about that!

There's https://github.com/llvm/llvm-project/pull/97856, https://github.com/llvm/llvm-project/pull/97861, and https://github.com/llvm/llvm-project/pull/97872 (still a draft).

I want to try and land some test infrastructure first (https://github.com/llvm/llvm-project/pull/101927) so I can ensure the refactorings don't break anything though. There are some more patches that I need to finish doing, but haven't gotten to the rest yet given the current ones are stalled.

**Comment by ganeshgit** [comment]
> This patch enables the basic skeleton enablement of AMD next gen zen5 CPUs.

@RKSimon Pl post your comments. I have few subsequent patches for scheduler enablement, and some tuning patches lined up as well.

**Review by RKSimon — APPROVED** [review]
LGTM as a base patch (znver4 + extra isas) - we should hold off from cherry picking into 19.x until we see the scope of the follow up patches.

**Comment by ganeshgit** [comment]
> @ganeshgit Can you address the clang-format warnings please?

Yes sure. I will correct them!

**Comment by RKSimon** [comment]
@tru This patch at the very least needs to make it for 19.x but I was hoping we'd get some of the tuning improvements in as well - should we wait for those PRs or just get this committed and cherry picked straight away?

**Comment by tru** [comment]
Hi,

This looks pretty safe, but feel free to think out loud about the risks of merging this. Technically it doesn't really fit the criteria of regression or important bugfix, I understand the needs of getting this into the release and I don't want to be a blocker for that.

Tuning improvements could be riskier if I understand it correctly?

**Comment by RKSimon** [comment]
@ganeshgit Ignore what I said earlier about waiting for the tuning patches :) Please can we get this committed to trunk, we'll let it brew for a few days and then cherry pick for 19.x - if you can create PRs for the tuning changes as soon as possible we can review them for 19.x on a case by case basis.

**Comment by nikic** [comment]
Zen 5 support in GCC was upstreamed more than half a year ago -- why is the LLVM support being upstreamed only now, after missing the 19.x window? What steps are being taken to ensure this does not happen again?

The change to X86TargetParser.h looks ABI breaking to me.

**Comment by ganeshgit** [comment]
> @ganeshgit Ignore what I said earlier about waiting for the tuning patches :) Please can we get this committed to trunk, we'll let it brew for a few days and then cherry pick for 19.x - if you can create PRs for the tuning changes as soon as possible we can review them for 19.x on a case by case basis.

I will upload the rebased code addressing the format errors shortly. Yes PRs will work for tuning changes. I think it will take some time.

**Comment by tru** [comment]
> The change to X86TargetParser.h looks ABI breaking to me.

This seems unfortunate to me. But I don't think it would be good to insert the enum at the end of the list and changing the sorting order.

How big of a problem would it be with a ABI break now? I know you have requested that we try to avoid those, even if it's strictly within the policy to still do that before the final release.

**Comment by RKSimon** [comment]
It would be messy, but could we not place the CK_ZNVER5 enum entry at the end of the enum list just for 19.x and then fix the sorting in trunk?

**Comment by AaronBallman** [comment]
> It would be messy, but could we not place the CK_ZNVER5 enum entry at the end of the enum list just for 19.x and then fix the sorting in trunk?

Seems better than an ABI break this late in the cycle, but I don't have *super* strong feelings.

**Comment by tru** [comment]
> It would be messy, but could we not place the CK_ZNVER5 enum entry at the end of the enum list just for 19.x and then fix the sorting in trunk?

I would be fine with that. WDYT @nikic ?

**Comment by tru** [comment]
I think it's better it land as it should be in `main` and then you can create a new PR against the ` release/19.x` branch with the different enum layout.

**Review by RKSimon — APPROVED** [review]

**Comment by RKSimon** [comment]
@ganeshgit OK to commit?

**Comment by ganeshgit** [comment]
> @ganeshgit OK to commit?

Yes I am okay. Can you please commit and close this PR. I will submit another PR for 19.x release.

**Comment by tru** [comment]
@ganeshgit or @RKSimon can one of you put up a PR against release/19.x with the abi compatible changes so that we have chance to look at it and approve it before the release tomorrow.

**Comment by ganeshgit** [comment]
> @ganeshgit or @RKSimon can one of you put up a PR against release/19.x with the abi compatible changes so that we have chance to look at it and approve it before the release tomorrow.

I will submit the changes shortly. Probably I will create a branch with a base commit which is prior to this commit to main for easy integration.

**Comment by tru** [comment]
Create a branch from `release/19.x` in your own fork, then cherry-pick over the changes from `main`, edit them to match the things we talked about above (and fix any merge problems). Then submit a PR that wants to merge `yourbranch` into `release/19.x`. And check the checkbox `maintainers can edit this branch` and it will help my integration!

**Pull request merged**
